### PR TITLE
[jit] Support custom error messages for `raise`

### PR DESCRIPTION
**Pull request by driazati**
Stacked PRs
 * **#34112 - [jit] Support custom error messages for `raise`**

Adds support for custom error messages in TorchScript

* Previously we weren't emitting anything for the statement following `raise`, so this could break user code if there is un-scriptable stuff there
    * The torchvision and `torch.nn` tests all pass so I think this is OK to release
    * If someone has an unscriptable error message, they can use `torch.jit.is_scripting` to work around
* All exceptions eventually end up as a `RaiseException(str message)` op
    * As a hack, if the value passed to the exception is not a `str`, a conversion is added. This lets things like `raise ValueError(2)` look like they work

```python
@torch.jit.script
def fn(x):
    raise RuntimeError("My error")
```

outputs

```
torch.jit.Error: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
  File "test.py", line 8, in fn
@torch.jit.script
def fn():
    raise RuntimeError("My error")
    ~~~~~~~~~~~~~~~~~~~~ <--- HERE
RuntimeError: My error
```

instead of the old

```
torch.jit.Error: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
  File "test.py", line 5, in fn
@torch.jit.script
def fn():
    raise RuntimeError("My error")
    ~~~~~~~~~~~~~~~~~~~~ <--- HERE
RuntimeError: Exception
```

Differential Revision: [D20496973](https://our.internmc.facebook.com/intern/diff/20496973/)

### CODE REVIEW

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 1]
```diff
@@ -1723,19 +1721,58 @@ struct to_ir {
   // raise a
   //
   // We ignore the expression following raise
-  void emitRaise(const SourceRange& loc) {
-    const std::string exception = "Exception";
-    auto string_input = insertConstant(*graph, exception, loc);
-    graph->insert(prim::RaiseException, {string_input}, {}, loc);
+  void emitRaise(const Raise& raise) {
+    auto sv = emitSugaredExpr(raise.expr(), 1);
```
Comment (eellison) [inline torch/csrc/jit/frontend/ir_emitter.cpp]: Can we move this all to python_sugared_values ? and put all the conversion of the inputs to strings in there.

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 2]
```diff
@@ -473,6 +488,9 @@ struct Environment {
           {"ord", std::make_shared<BuiltinFunction>(aten::ord, at::nullopt)},
           {"chr", std::make_shared<BuiltinFunction>(aten::chr, at::nullopt)},
           {"bin", std::make_shared<BuiltinFunction>(aten::bin, at::nullopt)},
+          {"AssertionError",
+           std::make_shared<ExceptionValue>("AssertionError")},
+          {"RuntimeError", std::make_shared<ExceptionValue>("RuntimeError")},
```
Comment (eellison) [inline torch/csrc/jit/frontend/ir_emitter.cpp]: Hi-level feedback: I think we want to move this stuff into python resolution, instead of special casing these two specific exceptions.

**Review by eellison — COMMENTED** [review]
Didn't do full review but my general comment is looks like the right approach. I think it would be good to move stuff out of the compiler and into python_sugared_values / python code. If from the compiler side it just gets raise {str} that would be nice. 

Can you double check that you need to add compiler builtins ? not sure what this would be needed. I think your approach of raising `{ExceptionName}: {str of inputs}` is good .

File: torch/jit/_builtins.py [hunk 3]
```diff
@@ -139,3 +139,49 @@ def _register_builtin(fn, op):
...
 def _find_builtin(fn):
     return _get_builtin_table().get(id(fn))
+
+
+# These are Python globals that have special sugaring in the compiler. See
+# [python globals] in the compiler for details. Adding a builtin global here
+# indicates that when resolving, we should skip these so they dispatch to
+# the correct sugared representation
+compiler_builtins = None
+def _is_global_builtin(maybe_builtin):
```
[the remark this answers is not in the document]
Comment (driazati) [reply torch/jit/_builtins.py]: Right now if someone does

```python
def list(x):
    return x + 1

@torch.jit.script
def fn(x):
    return list(x)
```

It will use `aten::list` instead of compiling the actual function since our resolution is not correct. I ran into this when I added `AssertionError` and `RuntimeError` as globals in the compiler (we need those to support `RuntimeError`s inside `cu.define` which is used in a couple places already), so this says that if we have a builtin Python function, don't resolve it with the resolver if it is the actual builtin.

This does add an extra step, but whenever someone adds a new compiler builtin it won't work until they add it here (since it will just resolve to the Python version and get put in as a Python call), so I think this is a pretty safe change

[the remark this answers is not in the document]
Comment (eellison) [reply torch/jit/_builtins.py]: I don't think we need to support `RuntimeError` inside `cu.define`. We don't generally support everything in string parsing that we do normally. See: https://github.com/pytorch/pytorch/issues/31391#issuecomment-567206434

[the remark this answers is not in the document]
Comment (driazati) [reply torch/jit/_builtins.py]: We already have code that uses it (I forget where exactly though) and the alternative to have a raise inside of a string is to do something like `raise "exception"`, which goes against the whole "its just a subset of Python" thing. It also makes things like `AssertionError` easier to implement since it just de-sugars `assert cond, msg` to an if statement with a `raise AssertionError(msg)`, but the compiler needs to know how to compile an exception to do that

[the remark this answers is not in the document]
Comment (eellison) [reply torch/jit/_builtins.py]: By code that uses it do you mean `checkScript` tests. There are 63 builtin exceptions. Why should we support `AssertionError` and `RuntimeError` but not `ValueError` or any of the other ones ? 

Other things we don't support with compilation unit: `torch.dtype`, list comprehensions, class types, etc. All of our existing files are serialized as `raise "exception"` so this is BC we have to support. Given that we need to support this it should be not very hard to support `assert cond, msg`.

[the remark this answers is not in the document]
Comment (driazati) [reply torch/jit/_builtins.py]: Going through it again I couldn't find any case of `raise RuntimeError` in a `.define`, so I'm not sure what was going on there, so we don't need `RuntimeError` as a compiler builtin.

But that doesn't change the general need for this since the compiler is incorrect today since it doesn't respect shadowing Python builtins. It's probably a rare case but I don't think this adds any huge tech debt since there's no way for this list to fall out of date with the one in the compiler.

Where are exceptions serialized as `raise "error"`? I can't find any reference to that in `python_print.cpp` and trying it out they get serialized as `ops.prim.RaiseException("error")`

[the remark this answers is not in the document]
Comment (eellison) [reply torch/jit/_builtins.py]: Okay, looks like i was wrong on how its serialized. 

Shadowing of Python builtins is a real problem but it would make this easier to review if that PR was separated from this with this PR. Potentially via ghstack. The only values that we need as compiler builtins are ones that we need to support serialized models. "list" only ever gets serialized as `torch.ops.list`, so it doesn't need to be a compiler builtin. It could instead be registered in `_builtins` as `Iid(list), aten::list)`.

[the remark this answers is not in the document]
Comment (driazati) [reply torch/jit/_builtins.py]: That and there are a few other issues with these builtins (e.g. #34851) that we could do as a follow up. Agreed it probably should have been a stack but since this depends on the builtin resolution for the changes to asserts the changes are pretty tangled up now .

[the remark this answers is not in the document]
Comment (eellison) [reply torch/jit/_builtins.py]: You're making a lot of changes to both compiler internals and one of the more tricky parts of the code base (sugared values / python sugared values), i think it's a reasonable ask to separate to two PRs.

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 4]
```diff
@@ -494,6 +494,8 @@ struct Environment {
           {"ord", std::make_shared<BuiltinFunction>(aten::ord, at::nullopt)},
           {"chr", std::make_shared<BuiltinFunction>(aten::chr, at::nullopt)},
           {"bin", std::make_shared<BuiltinFunction>(aten::bin, at::nullopt)},
+          {"AssertionError",
```
Comment (eellison) [inline torch/csrc/jit/frontend/ir_emitter.cpp]: I think we should only bind `Exception` which is the base class that has 62 other classes that inherit from it.

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 5]
```diff
@@ -1724,19 +1726,58 @@ struct to_ir {
...
   // We ignore the expression following raise
-  void emitRaise(const SourceRange& loc) {
-    const std::string exception = "Exception";
-    auto string_input = insertConstant(*graph, exception, loc);
-    graph->insert(prim::RaiseException, {string_input}, {}, loc);
+  void emitRaise(const Raise& raise) {
+    auto sv = emitSugaredExpr(raise.expr(), 1);
+    Value* value = nullptr;
+    auto exception_sv = std::dynamic_pointer_cast<ExceptionMessageValue>(sv);
+    if (exception_sv == nullptr) {
```
Comment (eellison) [inline torch/csrc/jit/frontend/ir_emitter.cpp]: The nesting here is a little hard to follow, and with the `if (value == nullptr)` and the `exception_sv` below. Could you try to rewrite / refactor ? maybe by consolidating ExceptionValue & ExceptionMessageValue or having one inherit from the other

File: torch/csrc/jit/frontend/sugared_value.h [hunk 6]
```diff
@@ -651,5 +651,44 @@ struct SimpleSelf : public Self {
...
+  Value* getValue() {
+    return value_;
+  }
+
+  Value* value_;
+};
+
+
+struct TORCH_API ExceptionValue : public SugaredValue {
+  ExceptionValue(const std::string& message) : message_(std::move(message)) {}
```
Comment (eellison) [inline torch/csrc/jit/frontend/sugared_value.h]: Do we need both `ExceptionValue` and `ExceptionMessageValue` ? or could one extend from the other ?

File: torch/csrc/jit/python/python_sugared_value.cpp [hunk 7]
```diff
@@ -548,12 +548,29 @@ std::shared_ptr<SugaredValue> BooleanDispatchValue::call(
...
+    at::ArrayRef<NamedValue> inputs,
+    at::ArrayRef<NamedValue> attributes,
+    size_t n_binders) {
+  Value* message;
+  if (inputs.size() == 0) {
+    message = insertConstant(*caller.graph(), "", loc);
+  } else if (inputs.size() == 1) {
+    message = inputs.at(0).value(*caller.graph());
+  } else {
+    throw ErrorReport(loc) << "exceptions in TorchScript only support"
```
Comment (eellison) [inline torch/csrc/jit/python/python_sugared_value.cpp]: Why not construct a tuple from the inputs ?

File: torch/jit/__init__.py [hunk 8]
```diff
@@ -1094,6 +1094,12 @@ def _is_new_style_class(cls):
         return ('__dict__' in dir(cls) or hasattr(cls, '__slots__'))
 
 
+def _is_exception(obj):
+    if not inspect.isclass(obj):
```
Comment (eellison) [inline torch/jit/__init__.py]: Do we need `if not inspect.isclass(obj):` ?

**Review by eellison — COMMENTED** [review]
Few more comments

File: torch/jit/__init__.py (hunk 8 above)
Comment (driazati) [reply torch/jit/__init__.py]: `issubclass` requires a class as the second argument

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 4 above)
Comment (driazati) [reply torch/csrc/jit/frontend/ir_emitter.cpp]: `AssertionError` is special here since we don't have real assert statements, we don't really need to add any others. This PR changes them into, so the compiler needs to know about `AssertionError`:
```python
if not cond:
    raise AssertionError(message)
```

File: torch/csrc/jit/frontend/sugared_value.h (hunk 6 above)
Comment (driazati) [reply torch/csrc/jit/frontend/sugared_value.h]: `ExceptionValue` corresponds to `RuntimeError` and `ExceptionMessageValue` corresponds to the instance returned by `RuntimeError("error")`, so we do need both since it is valid to do something like `raise RuntimeError` and `raise RuntimeError("error")`

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 9]
```diff
@@ -1728,19 +1730,58 @@ struct to_ir {
   // raise a
   //
   // We ignore the expression following raise
-  void emitRaise(const SourceRange& loc) {
-    const std::string exception = "Exception";
-    auto string_input = insertConstant(*graph, exception, loc);
-    graph->insert(prim::RaiseException, {string_input}, {}, loc);
+  void emitRaise(const Raise& raise) {
+    auto sv = emitSugaredExpr(raise.expr(), 1);
+    Value* value = nullptr;
```
Comment (eellison) [inline torch/csrc/jit/frontend/ir_emitter.cpp]: nit: `s/value/error_msg`

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 10]
```diff
@@ -1728,19 +1730,58 @@ struct to_ir {
...
+      //    raise RuntimeError
+      value = insertConstant(*graph, "", raise.range());
+    } else {
+      // The raise was not followed by an exception (i.e. it was something like
+      // `raise "error"` instead of `raise RuntimeError("error")`)
+      throw ErrorReport(raise.range())
+          << "exceptions must derive from BaseException";
+    }
+
+    // TODO: Fix hack (some things like `ValueError` can take non-string
```
Comment (eellison) [inline torch/csrc/jit/frontend/ir_emitter.cpp]: I don't think this is should be here...

File: torch/csrc/jit/python/python_sugared_value.cpp (hunk 7 above)
Comment (eellison) [reply torch/csrc/jit/python/python_sugared_value.cpp]: This isn't resolved. I think an exception that has multiple inputs will be a very common case.

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 11]
```diff
@@ -1728,19 +1730,58 @@ struct to_ir {
...
+    // arguments) Since exceptions are still sugared values and don't have an
+    // `attr()` defined, this should be ok for now
+    if (!value->type()->isSubtypeOf(StringType::get())) {
+      value = graph->insert(aten::str, {value});
```
Comment (eellison) [inline torch/csrc/jit/frontend/ir_emitter.cpp]: This is going to fail at runtime instead of compilation...

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 12]
```diff
@@ -1728,19 +1730,58 @@ struct to_ir {
...
+  void emitRaise(const Raise& raise) {
+    auto sv = emitSugaredExpr(raise.expr(), 1);
+    Value* value = nullptr;
+
+    auto exception_instance = std::dynamic_pointer_cast<ExceptionMessageValue>(sv);
+    if (auto exception_instance =
+            std::dynamic_pointer_cast<ExceptionMessageValue>(sv)) {
+      // The typical case, an instance of the exception class was thrown:
+    //    raise RuntimeError("error")
+      value = exception_instance->getValue();
```
Comment (eellison) [inline torch/csrc/jit/frontend/ir_emitter.cpp]: Can we: have `ExceptionMessageValue` store all of the values it is called with as a list of values. Then, here, try to cast each one to str ? ideally with `str` cast value above.

**Review by eellison — COMMENTED** [review]
couple more comments i think we want to address before landing.

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 11 above)
Comment (driazati) [reply torch/csrc/jit/frontend/ir_emitter.cpp]: How would it fail at runtime? If it doesn't work there will be a schema-mismatch

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 12 above)
Comment (driazati) [reply torch/csrc/jit/frontend/ir_emitter.cpp]: This is now handled by making the sugared value construct a tuple

File: torch/_jit_internal.py [hunk 13]
```diff
@@ -84,6 +84,8 @@ def __getattr__(self, key):
                 return f_locals[key]
             elif key in f_globals:
                 return f_globals[key]
+            elif key in dir(builtins):
+                return getattr(builtins, key)
```
[the remark this answers is not in the document]
Comment (driazati) [reply torch/_jit_internal.py]: The stack-based resolution callback doesn't include globals (while the closure based one does), this just makes them have the same functionality (so `ValueError` can be resolved)

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 4 above)
Comment (eellison) [reply torch/csrc/jit/frontend/ir_emitter.cpp]: What do you think ?

File: torch/csrc/jit/python/python_sugared_value.h [hunk 14]
```diff
@@ -266,5 +266,25 @@ struct VISIBILITY_HIDDEN PythonClassValue : public ClassValue {
...
+  }
+
+  std::shared_ptr<SugaredValue> call(
+      const SourceRange& loc,
+      Function& caller,
+      at::ArrayRef<NamedValue> inputs,
+      at::ArrayRef<NamedValue> attributes,
+      size_t n_binders) override;
+
+  py::object exception_class_;
```
Comment (eellison) [inline torch/csrc/jit/python/python_sugared_value.h]: This is never used.

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 15]
```diff
@@ -1728,19 +1730,54 @@ struct to_ir {
...
+    if (auto exception_instance =
+            std::dynamic_pointer_cast<ExceptionMessageValue>(sv)) {
+      // The typical case, an instance of the exception class was thrown:
+      //    raise RuntimeError("error")
+      error_message = exception_instance->getValue();
+    } else if (
+        auto exception_class = std::dynamic_pointer_cast<ExceptionValue>(sv)) {
+      // A bare exception was thrown so add an empty message. e.g.
+      //    raise RuntimeError
+      error_message = insertConstant(*graph, "", raise.range());
```
Comment (eellison) [inline torch/csrc/jit/frontend/ir_emitter.cpp]: reason why don't make the error message the name of the exception ?

**Review by eellison — COMMENTED** [review]
This looks good but I'm not sure you have any tests for the `raise RuntimeError` case, so i'm not sure if that works.

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 11 above)
Comment (eellison) [reply torch/csrc/jit/frontend/ir_emitter.cpp]: Yea i take this back, insert is the api that checks schemas

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 15 above)
Comment (driazati) [reply torch/csrc/jit/frontend/ir_emitter.cpp]: It'd end up like `RuntimeError: ValueError` which is a weird in between state. As a follow up this should hook into Python and throw the right exception class.

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 4 above)
Comment (driazati) [reply torch/csrc/jit/frontend/ir_emitter.cpp]: Added a comment, why should we bind in `Exception`? If its for things like AD formulas I think we can just wait and see what is specifically requested before adding it

**Comment by driazati** [comment]
The builtins resolution stuff isn't necessary anymore, so this should be ready

File: torch/csrc/jit/python/python_sugared_value.cpp [hunk 16]
```diff
@@ -548,12 +548,34 @@ std::shared_ptr<SugaredValue> BooleanDispatchValue::call(
   return value->call(loc, caller, inputs, attributes, n_binders);
 }
 
+std::shared_ptr<SugaredValue> PythonExceptionValue::call(
+    const SourceRange& loc,
+    Function& caller,
+    at::ArrayRef<NamedValue> inputs,
+    at::ArrayRef<NamedValue> attributes,
```
Comment (eellison) [inline torch/csrc/jit/python/python_sugared_value.cpp]: We should either not ignore attributes or throw if they're present

File: test/test_jit.py [hunk 17]
```diff
@@ -16619,7 +16629,7 @@ def test_if_might(x):
                 else:
                     a = 2
             else:
-                raise RunTimeError("hi")
+                raise RuntimeError("hi")
```
Comment (eellison) [inline test/test_jit.py]: Can you add a test case for :`raise RuntimeError` (no msg)

**Review by eellison — APPROVED** [review]
Ship it!
